src/xml/writer.cr

```diff
@@ -0,0 +1,312 @@
+        ->(ctx) {
+          Box(IO).unbox(ctx).flush
+          0
+        },
+        @box,
+        nil
+      )
+      @writer = LibXML.xmlNewTextWriter(buffer)
+    end
+
+    # Emits the start of the document.
+    def start_document(version = nil, encoding = nil) : Nil
+      call StartDocument, version, encoding, nil
+    end
+
+    # Emits the end of a document.
+    def end_document : Nil
+      call EndDocument
+    end
+
+    # Emits the start of the document, invokes the block,
+    # and the emits the end of the document.
+    def document(version = nil, encoding = nil)
+      start_document version, encoding
+      yield.tap { end_document }
+    end
+
+    # Emits the start of an element.
+    def start_element(name : String) : Nil
+      call StartElement, name
+    end
+
+    # Emits the start of an element with namespace info.
+    def start_element(prefix : String?, name : String, namesapce_uri : String?) : Nil
+      call StartElementNS, string_to_unsafe(prefix), name, string_to_unsafe(namesapce_uri)
+    end
+
+    # Emits the end of an element.
+    def end_element : Nil
+      call EndElement
+    end
+
+    # Emits the start of an element with the given attributes,
+    # invokes the block and then emits the end of the element.
+    def element(__name__ : String, **attributes)
+      element(__name__, attributes) do
+        yield
+      end
+    end
+
+    # ditto
+    def element(__name__ : String, attributes : Hash | NamedTuple)
+      start_element __name__
+      attributes(attributes)
+      yield.tap { end_element }
+    end
+
+    # Emits an element with the given attributes.
+    def element(__name__ : String, **attributes)
+      element(__name__, attributes)
+    end
+
+    # ditto
+    def element(name : String, attributes : Hash | NamedTuple)
+      element(name, attributes) { }
+    end
+
+    # Emits the start of an element with namespace info with the given attributes,
+    # invokes the block and then emits the end of the element.
+    def element(__prefix__ : String?, __name__ : String, __namespace_uri__ : String?, **attributes)
+      element(__prefix__, __name__, __namespace_uri__, attributes) do
+        yield
+      end
+    end
+
+    # ditto
+    def element(__prefix__ : String?, __name__ : String, __namespace_uri__ : String?, attributes : Hash | NamedTuple)
+      start_element __prefix__, __name__, __namespace_uri__
+      attributes(attributes)
+      yield.tap { end_element }
+    end
+
+    # Emits an element with namespace info with the given attributes.
+    def element(prefix : String?, name : String, namespace_uri : String?, **attributes)
+      element(prefix, name, namespace_uri, attributes)
+    end
+
+    # ditto
+    def element(prefix : String?, name : String, namespace_uri : String?, attributes : Hash | NamedTuple)
+      start_element(prefix, name, namespace_uri)
+      attributes(attributes)
+      end_element
+    end
+
+    # Emits the start attribute of an attribute.
```

Comment:
The "start attribute of an attribute"? Did you mean the "start of an attribute"? Same below.